go/tools/builders/pack.go

```diff
@@ -80,19 +84,182 @@ func copyFile(inPath, outPath string) error {
 	return err
 }
 
-func listFiles(ar, archive string) ([]string, error) {
-	cmd := exec.Command(ar, "t", archive)
-	out, err := cmd.Output()
+const (
+	// arHeader appears at the beginning of archives created by "ar" and
+	// "go tool pack" on all platforms.
+	arHeader = "!<arch>\n"
+
+	// entryLength is the size in bytes of the metadata preceding each file
+	// in an archive.
+	entryLength = 60
+)
+
+func extractFiles(archive string) (files []string, err error) {
+	f, err := os.Open(archive)
 	if err != nil {
 		return nil, err
 	}
-	lines := strings.Split(string(out), "\n")
-	var files []string
-	for _, line := range lines {
-		line = strings.TrimSpace(line)
-		if line != "" {
-			files = append(files, line)
+	defer f.Close()
+	r := bufio.NewReader(f)
+
+	header := make([]byte, len(arHeader))
+	if _, err := io.ReadFull(r, header); err != nil || string(header) != arHeader {
+		return nil, fmt.Errorf("%s: bad header", archive)
+	}
+
+	var nameData []byte
+	for {
+		name, size, err := readMetadata(r, &nameData)
+		if err == io.EOF {
+			return files, nil
+		}
+		if err != nil {
+			return nil, err
 		}
+		if !isObjectFile(name) {
+			if err := skipFile(r, size); err != nil {
+				return nil, err
+			}
+			continue
+		}
+		if err := extractFile(r, name, size); err != nil {
+			return nil, err
+		}
+		files = append(files, name)
+	}
+}
+
+// readMetadata reads the relevant fields of an entry. Before calling,
+// r must be positioned at the beginning of an entry. Afterward, r will
+// be positioned at the beginning of the file data. io.EOF is returned if
+// there are no more files in the archive.
+//
+// Both BSD and GNU / SysV naming conventions are supported.
+func readMetadata(r *bufio.Reader, nameData *[]byte) (name string, size int64, err error) {
+retry:
+	// Each file is preceded by a 60-byte header that contains its metadata.
+	// We only care about two fields, name and size. Other fields (mtime,
+	// owner, group, mode) are ignored because they don't affect compilation.
+	var entry [entryLength]byte
+	if _, err := io.ReadFull(r, entry[:]); err != nil {
```

Comment:
You probably need to process the entry anyway if it was read in full but returned an eof

Reply:
`ReadFull` should work, even though it drops other errors.